beacon-chain/sync/subscriber_beacon_blocks_chunks.go

```diff
@@ -0,0 +1,186 @@
+		return pubsub.ValidationIgnore, nil
+	} else if err != nil {
+		// TODO: cook up a slashing object if the error is ErrIncorrectCommitments
+		log.WithError(err).Debug("Could not validate beacon block chunk")
+		return pubsub.ValidationReject, err
+	}
+	logFields := logrus.Fields{
+		"chunkSlot":     chunk.Slot(),
+		"proposerIndex": chunk.ProposerIndex(),
+		"parentRoot":    chunk.ParentRoot(),
+	}
+	log.WithFields(logFields).Debug("Received block chunk")
+
+	// If the block can be recovered, send it to the blockchain package
+	go s.reconstructBlockFromChunk(ctx, chunk)
+	go s.broadcastBlockChunk(ctx, chunk)
+	return pubsub.ValidationIgnore, nil
+}
+
+func (s *Service) validateBeaconBlockChunk(ctx context.Context, chunk interfaces.ReadOnlyBeaconBlockChunk) error {
+	if !s.cfg.chain.InForkchoice(chunk.ParentRoot()) {
+		return blockchain.ErrNotDescendantOfFinalized
+	}
+	err := s.blockChunkCache.AddChunk(chunk)
+	if err == nil {
+		return nil
+	}
+	if errors.Is(err, rlnc.ErrSignatureNotVerified) {
+		parentState, err := s.cfg.stateGen.StateByRoot(ctx, chunk.ParentRoot())
+		if err != nil {
+			s.blockChunkCache.RemoveNode(chunk) // Node is guaranteed to have a single chunk
+			return err
+		}
+		if err := core_chunks.VerifyChunkSignatureUsingCurrentFork(parentState, chunk); err != nil {
+			s.blockChunkCache.RemoveNode(chunk) // Node is guaranteed to have a single chunk
+			return err
+		}
+		return nil
+	}
+	return err
+}
+
+// startChunkPruner starts a goroutine that prunes the block chunk cache every epoch.
+func (s *Service) startChunkPruner() {
+	cp := s.cfg.chain.FinalizedCheckpt()
+	fSlot, err := slots.EpochStart(cp.Epoch)
+	if err != nil {
+		log.WithError(err).Debug("could not prune the chunk cache: could not calculate epoch start slot")
+	} else {
+		s.blockChunkCache.Prune(fSlot)
+	}
+}
+
+func (s *Service) reconstructBlockFromChunk(ctx context.Context, chunk interfaces.ReadOnlyBeaconBlockChunk) {
+	data, err := s.blockChunkCache.GetBlockData(chunk.Slot(), chunk.ProposerIndex())
+	if err != nil {
+		return
+	}
+
+	msg := p2p.GossipTopicMappings(p2p.BlockSubnetTopicFormat, slots.ToEpoch(chunk.Slot()))
+	e := &encoder.SszNetworkEncoder{}
+	msgSSZ, ok := msg.(fastssz.Unmarshaler)
+	if !ok {
+		logrus.Error("Could not convert message to fastssz.Unmarshaler")
+		return
+	}
+	if err := e.DecodeGossip(data, msgSSZ); err != nil {
+		logrus.WithError(err).Error("Could not decode block data")
+		return
+	}
+	if err := s.beaconBlockSubscriber(ctx, msg); err != nil {
+		logrus.WithError(err).Error("Could not handle p2p pubsub")
+	}
+}
+
+func (s *Service) broadcastBlockChunk(ctx context.Context, chunk interfaces.ReadOnlyBeaconBlockChunk) {
+	msg, err := s.blockChunkCache.PrepareMessage(chunk)
+	if err != nil {
+		logrus.WithError(err).Error("could not broadcast chunk")
+		return
+	}
+	if err := s.cfg.p2p.Broadcast(ctx, msg); err != nil {
+		logrus.WithError(err).Error("could not broadcast chunk")
+	}
```

Comment:
This is wrong, instead of broadcasting the node should send a different chunk to each one of its peers.